### PR TITLE
chore: add typed resource

### PR DESCRIPTION
**Pull request by DmitriyMV**
Add TypedResource type and tests for it

Signed-off-by: Dmitriy Matrenichev <dmitry.matrenichev@siderolabs.com>

### CODE REVIEW

File: pkg/resource/typed/typed_resource.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+// DeepCopyable requires a spec to have DeepCopy method which will be used during TypedResource copy.
+type DeepCopyable[T any] interface {
+	DeepCopy() T
+}
+
+// ResourceDefinition is a phantom type which acts as info supplier for TypedResource String and ResourceDefinition
+// methods. It intantianed only during String and ResourceDefinition calls, so it should never contain any data which
+// survies those calls.
```
Comment (frezbo) [inline pkg/resource/typed/typed_resource.go:19]:
```suggestion
// survives those calls.
```

Comment (DmitriyMV) [reply pkg/resource/typed/typed_resource.go]: Thx! done

File: pkg/resource/typed/types_resource_test.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/resource/meta"
+	"github.com/cosi-project/runtime/pkg/resource/typed"
+)
+
+type TestSpec struct {
```
[the remark this answers is not in the document]
Comment (DmitriyMV) [reply pkg/resource/typed/types_resource_test.go:17]: Done!

File: pkg/resource/typed/types_resource_test.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	Var int
+}
+
+func (t TestSpec) DeepCopy() TestSpec {
+	return t
+}
+
+type Test = typed.TypedResource[TestSpec, TestRD]
```
Comment (smira) [inline pkg/resource/typed/types_resource_test.go:25]: this feels like it should be `typed.Resource`

File: pkg/resource/typed/types_resource_test.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	resCopy := res.DeepCopy()
+	asrt.True(resCopy.Metadata().Equal(*res.Metadata()))
+	checkResourceEqualSpec(t, resCopy, res)
+
+	// check that modifying one doesn't modify the other
+	res.TypedSpec().Var = 45
+	asrt.NotEqual(resCopy.Spec(), res.Spec())
+
+	// check that getting resource definition actually works on phantom types
+	asrt.Equal(res.ResourceDefinition().DisplayType, "test definition")
```
Comment (smira) [inline pkg/resource/typed/types_resource_test.go:77]: I think we need `asrt.Implements` that `*Test` implements a `resource.Resource`

Comment (DmitriyMV) [reply pkg/resource/typed/types_resource_test.go:68]: Done using:
```
var _ resource.Resource = (*Test)(nil)
```

File: pkg/resource/typed/types_resource_test.go (hunk 3 above)
Comment (DmitriyMV) [reply pkg/resource/typed/types_resource_test.go]: Done!

**Comment by DmitriyMV** [comment]
Okay, what I do not like, is that if you try to do `make generate fmt lint` lint stops being okay with
```
#30 0.872 pkg/state/impl/inmem/build.go:13:53: directive `//nolint:ireturn` is unused for linter "ireturn" (nolintlint)
#30 0.872 func Build(ns resource.Namespace) state.CoreState { //nolint:ireturn
#30 0.872                                                     ^
#30 0.872 pkg/state/impl/inmem/collection.go:60:96: directive `//nolint:ireturn` is unused for linter "ireturn" (nolintlint)
#30 0.872 func (collection *ResourceCollection) Get(resourceID resource.ID) (resource.Resource, error) { //nolint:ireturn
#30 0.872                                                                                                ^
#30 0.872 pkg/state/impl/inmem/inmem.go:57:131: directive `//nolint:ireturn` is unused for linter "ireturn" (nolintlint)
#30 0.872 func (st *State) Get(ctx context.Context, resourcePointer resource.Pointer, opts ...state.GetOption) (resource.Resource, error) { //nolint:ireturn
#30 0.872                                                                                                                                   ^
#30 ERROR: process "/bin/sh -c golangci-lint run --config .golangci.yml" did not complete successfully: exit code: 1
```

What is funny - thats not the files we are regenerating.

**Review by smira — APPROVED** [review]
:nerd_face:

**Pull request merged**